### PR TITLE
unCLIP image variation

### PR DESCRIPTION
**Pull request by williamberman**
Adds an unclip image variation pipeline

#### Converting the text to image pipeline to image variation

I uploaded the pipeline to https://huggingface.co/fusing/karlo-image-variations-diffusers if you want to skip this

From the diffusers root directory:
```bash
$ python scripts/convert_unclip_txt2img_to_image_variation.py --dump_path <path to save model>
```

#### Using the model

```python
import os
os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"

from diffusers import UnCLIPImageVariationPipeline
import torch
import random
import numpy as np
import PIL
from PIL import Image


def set_seed(seed: int):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
set_seed(0)

torch.backends.cuda.matmul.allow_tf32 = False
torch.use_deterministic_algorithms(True)
torch.set_printoptions(precision=40)

def image_grid(imgs, rows, cols):
    assert len(imgs) == rows*cols

    w, h = imgs[0].size
    grid = Image.new('RGB', size=(cols*w, rows*h))
    
    for i, img in enumerate(imgs):
        grid.paste(img, box=(i%cols*w, i//cols*h))
    return grid


pipe = UnCLIPImageVariationPipeline.from_pretrained("fusing/karlo-image-variations-diffusers")
pipe = pipe.to('cuda')

# See image below to use as input
image = PIL.Image.open("./test.jpg") 

images = pipe(image, num_images_per_prompt=4).images

image_grid(images, 1, 4).save('./out.jpg')
```

test

out

### CODE REVIEW

File: src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+                latents = torch.randn(shape, generator=generator, device=device, dtype=dtype)
+        else:
+            if latents.shape != shape:
+                raise ValueError(f"Unexpected latents shape, got {latents.shape}, expected {shape}")
+            latents = latents.to(device)
+
+        latents = latents * scheduler.init_noise_sigma
+        return latents
+
+    def _encode_prompt(self, prompt, device, num_images_per_prompt, do_classifier_free_guidance):
```
Comment (patil-suraj) [inline src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py:128]: Think we can add `Copied from ...` comments here, this is same as in unclip text2image pipeline no ?

File: src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+                Whether or not to return a [`~pipeline_utils.ImagePipelineOutput`] instead of a plain tuple.
+        """
+        if isinstance(image, PIL.Image.Image):
+            batch_size = 1
+        elif isinstance(image, list):
+            batch_size = len(image)
+        else:
+            batch_size = image.shape[0]
+
+        prompt = [""] * batch_size
```
Comment (patil-suraj) [inline src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py:313]: why do we need an empty prompt here?

**Review by patil-suraj — APPROVED** [review]
Looks good to me, thanks a lot for adding the pipeline! Would be nice to add `copied from ..` comments wherever possible.

File: src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py (hunk 2 above)
Comment (williamberman) [reply src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py:313]: The original image variation pipeline still uses the conditional unet just with an empty prompt https://github.com/kakaobrain/karlo/blob/0b5dbb0c9f34363921a5c3bf2a78dfe8e3ad6ceb/karlo/sampler/i2i.py#L113

File: src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py (hunk 1 above)
Comment (williamberman) [reply src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py:128]: hmm, I didn't add the copied from comments in the text to image unclip pipeline. I'm not sure if there are standards on how we use those. Is there a canonical pipeline that we "copy from"?

File: src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py [hunk 3]
```diff
@@ -0,0 +1,454 @@
+# Copyright 2022 Kakao Brain and The HuggingFace Team. All rights reserved.
```
Comment (pcuenca) [inline src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py:1]: Is this so? Or is it just Hugging Face for the code? Just wondering, no idea how those things work!

File: src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+    Args:
+        text_encoder ([`CLIPTextModelWithProjection`]):
+            Frozen text-encoder.
+        tokenizer (`CLIPTokenizer`):
+            Tokenizer of class
+            [CLIPTokenizer](https://huggingface.co/docs/transformers/v4.21.0/en/model_doc/clip#transformers.CLIPTokenizer).
+        feature_extractor ([`CLIPFeatureExtractor`]):
+            Model that extracts features from generated images to be used as inputs for the `image_encoder`.
+        image_encoder ([`CLIPVisionModelWithProjection`]):
+            Frozen CLIP image-encoder. Stable Diffusion Image Variation uses the vision portion of
```
Comment (pcuenca) [inline src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py:55]:
```suggestion
            Frozen CLIP image-encoder. unCLIP Image Variation uses the vision portion of
```

File: src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+        text_proj ([`UnCLIPTextProjModel`]):
+            Utility class to prepare and combine the embeddings before they are passed to the decoder.
+        decoder ([`UNet2DConditionModel`]):
+            The decoder to invert the image embedding into an image.
+        super_res_first ([`UNet2DModel`]):
+            Super resolution unet. Used in all but the last step of the super resolution diffusion process.
+        super_res_last ([`UNet2DModel`]):
+            Super resolution unet. Used in the last step of the super resolution diffusion process.
+        decoder_scheduler ([`UnCLIPScheduler`]):
+            Scheduler used in the decoder denoising process. Just a modified DDPMScheduler.
```
Comment (pcuenca) [inline src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py:67]: Does it need to be a DDPMScheduler? Wouldn't it work with DPM solver, for instance?

File: src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+        when their specific submodule has its `forward` method called.
+        """
+        if is_accelerate_available():
+            from accelerate import cpu_offload
+        else:
+            raise ImportError("Please install accelerate via `pip install accelerate`")
+
+        device = torch.device(f"cuda:{gpu_id}")
+
+        # TODO: self.prior.post_process_latents is not covered by the offload hooks, so it fails if added to the list
```
Comment (pcuenca) [inline src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py:226]:
```suggestion
```

Do we have `prior` here?

File: scripts/convert_unclip_txt2img_to_image_variation.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+    feature_extractor = CLIPImageProcessor()
+    image_encoder = CLIPVisionModelWithProjection.from_pretrained("openai/clip-vit-large-patch14")
+
+    img2img = UnCLIPImageVariationPipeline(
+        decoder=txt2img.decoder,
+        text_encoder=txt2img.text_encoder,
+        tokenizer=txt2img.tokenizer,
+        text_proj=txt2img.text_proj,
+        feature_extractor=feature_extractor,
```
Comment (pcuenca) [inline scripts/convert_unclip_txt2img_to_image_variation.py:32]: This is very informative, but I'm not sure we store this kind of scripts in the repo. The ones in the folder are usually about converting weights from other checkpoints. What do you think @patil-suraj?

**Review by pcuenca — APPROVED** [review]
Awesome!

File: src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py (hunk 3 above)
Comment (williamberman) [reply src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py:1]: This is the licensing @patrickvonplaten added to the text to image pipeline. We should probably clarify with him :)

File: src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py (hunk 5 above)
Comment (williamberman) [reply src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py:67]: Yeah so I would like to experiment with other schedulers when I get some time. The original codebase uses DDPM scheduling but with a few minor discrepancies with our normal DDPM implementation. We elected to make a separate unCLIP scheduler for the initial port into diffusers to assist with verifying against the karlo codebase.

File: src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py (hunk 6 above)
Comment (williamberman) [reply src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py:226]: Ah thanks for the catch!

File: scripts/convert_unclip_txt2img_to_image_variation.py (hunk 7 above)
Comment (williamberman) [reply scripts/convert_unclip_txt2img_to_image_variation.py:32]: Happy to remove and just put in the PR description! lmk @patil-suraj

File: src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py (hunk 4 above)
Comment (williamberman) [reply src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py:55]: Hehe 😓

File: src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py (hunk 1 above)
Comment (williamberman) [reply src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py]: looked like the unclip text to image pipeline had sufficient differences in its helper methods where none warranted a copy from. I added copy froms on the image variation pipeline helper methods :)

File: src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py (hunk 5 above)
Comment (pcuenca) [reply src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py:67]: Thanks for the clarification!

File: scripts/convert_unclip_txt2img_to_image_variation.py (hunk 7 above)
Comment (patil-suraj) [reply scripts/convert_unclip_txt2img_to_image_variation.py:32]: Yeah, think no need to have this script.

File: src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py (hunk 3 above)
Comment (patil-suraj) [reply src/diffusers/pipelines/unclip/pipeline_unclip_image_variation.py:1]: Think it's fine to mention Kakao Brain, since we use their code as a reference when implementing it here.

**Review by patil-suraj — APPROVED** [review]
LGTM! 
Also, think it would be nice to add a doc page explaining the unCLIP pipelines. It's the first cascaded pipeline in diffusers, so would be nice to document the different components and how they work.

**Review by patrickvonplaten — APPROVED** [review]

**Review by patrickvonplaten — COMMENTED** [review]
Nice looks good to me!

File: scripts/convert_unclip_txt2img_to_image_variation.py (hunk 7 above)
Comment (patrickvonplaten) [reply scripts/convert_unclip_txt2img_to_image_variation.py:32]: This is better than not having a script at all. Think it's totally fine to leave it here as is. The main purpose of the scripts is really so that the user can convert the checkpoints themselves - I'm fine with the way it is. Better would be to directly convert from the original checkpoint, but for me this is ok as well and def better than not having anything.
